---
title: "ASAP"
output: html_document
date: '2022-07-13'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)


library(tidyverse)
library(tidyr)

library(xml2)
```

```{r}



read.ASAP.snps <- function(XML){

   Out <- data.frame()
   
   xml_data <- read_xml(XML)
   
   Run_Info = cbind(Run = xml_attrs(xml_data))
   
   #for each sample create a node list with specific child
   for (SAMPLE in 1:xml_length(xml_data)) {
   
    #SAMPLE = 24
    #Subset xml_data
    Sample_Node <- xml_child(xml_data, SAMPLE) 
    
    #xml_attrs(Sample_Node) Get the attributes in the node
    
    #Pull out the attributes of interest
    Sample_Info <- cbind(
      Name = xml_attr(Sample_Node, "name")
      #Mapped_Reads = xml_attr(Sample_Node, "mapped_reads"),
      #unassigned_reads = xml_attr(Sample_Node, "unassigned_reads"),
      #unmapped_reads = xml_attr(Sample_Node, "unassigned_reads")
      )
    
      #For each assay within a sample
      for (ASSAY in 1:xml_length(Sample_Node)) {
      
        #ASSAY = 1
        #subset Sample node to assay
        Assay_Node <- xml_child(Sample_Node, ASSAY)
        
        #xml_attrs(Assay_Node) list the attrs for the assay
        
        #Pull out the attributes of interest
        Assay_Info <- cbind(
          assay_function = xml_attr(Assay_Node, "function"),
          gene = xml_attr(Assay_Node, "gene"),
          name = xml_attr(Assay_Node, "name"),
          type = xml_attr(Assay_Node, "type")
          )
        
        #For each amplicon in assay:
        for(AMPLICON in 1:xml_length(Assay_Node)){
          
          #AMPLICON = 1
          #Get a single amplicon node 
          Amplicon_Node <- xml_child(Assay_Node, AMPLICON)
         
          #xml_attrs(Amplicon_Node) #show the amplicon node attributes 
          
          Breadth = xml_contents(xml_child(Amplicon_Node, "breadth"))
          
          Amplicon_Info <- cbind(
              #Add the amplicon number to the amp_info
              Amplicon_Number = AMPLICON
              #Parse out number of reads in amplicon
              #Amplicon_reads = xml_attrs(Amplicon_Node, "reads"),
              #Parse out the breadth of coverage
              #Breadth = as.character(xml_contents(xml_child(Amplicon_Node, "breadth"))),
              #Parse out the average breadth
              #Avg_Depth = as.character(xml_contents(xml_child(Amplicon_Node, "average_depth")))
              )
      
          for (SNP in 1:xml_length(Amplicon_Node)) {
             #SNP = 1
             #Get a single amplicon node 
             SNP_Node <- xml_child(Amplicon_Node, SNP)
             
             if (xml_name(SNP_Node) == "snp"){
                
                xml_attrs(SNP_Node)
                
                xml_children(SNP_Node)
                
                if(xml_length(SNP_Node) == 2){
                  #Get and format the SNP_Distirbution Data 
                  SNP_Dist = as.character(xml_child(SNP_Node, 2))
                  SNP_Dist = str_remove(SNP_Dist, "<base_distribution ")
                  SNP_Dist = str_remove_all(SNP_Dist, "\"")
                  SNP_Dist = str_remove_all(SNP_Dist, "/>")
                }else{
                  SNP_Dist = "No reads matching SNP"
                }
                #Get and format SNP information
                SNP_call = as.character(xml_child(SNP_Node, 1))
                SNP_call = str_remove(SNP_call, "<snp_call count=")
                SNP_call = str_remove_all(SNP_call, "\"")
                SNP_call = str_remove_all(SNP_call, "percent=")
                SNP_call = str_replace(SNP_call, ">", " ")
                SNP_call = str_remove_all(SNP_call, "</snp_call>")
                SNP_call = as.list(str_split(SNP_call, " "))
                
                #Create object with values of interest
                SNP_Info <- cbind(
                   Loc_Depth = xml_attr(SNP_Node, "depth"),
                   SNP_Name = xml_attr(SNP_Node, "name"),
                   SNP_Position = xml_attr(SNP_Node, "position"),
                   SNP_Reference = xml_attr(SNP_Node, "reference"),
                   SNP_Depth = SNP_call[[1]][[1]],
                   SNP_Proportion = SNP_call[[1]][[2]],
                   SNP_Call = SNP_call[[1]][[3]],
                   SNP_Distribution = SNP_Dist
                )
               
                Temp <- cbind(Run_Info, Sample_Info, Assay_Info, Amplicon_Info, SNP_Info)
         
                Out <- rbind(Out, Temp)
            
                

             }

             
          }
          
              
         
        }
      }
   }
   
   row.names(Out) <- 1:nrow(Out)

   return(Out)
   
}


XML <- "/scratch/tporter/COVID_BAA_2022062122_Sanjeet_COH_Samples/Combined_BAA_MinorPop_NoSDSI_SMOR/Combined_BAA_MinorPop_NoSDSI_SMOR_analysis.xml"
SNPS <- read.ASAP.snps(XML)
ASAP <- read.ASAP(XML)

str_split(ASAP[8,15], ",")[[1]][[23048]]

```

